public/app/features/variables/adhoc/AdHocVariableEditor.tsx

```diff
@@ -38,8 +36,8 @@ export class AdHocVariableEditorUnConnected extends PureComponent<Props> {
 
   render() {
     const { variable, editor } = this.props;
-    const dataSources = editor.extended?.dataSources ?? [];
-    const infoText = editor.extended?.infoText ?? null;
+    const dataSources = (editor.extended?.kind === 'adhoc' && editor.extended?.dataSources) || [];
```

Comment:
Hmm, where does kind come from, the variable already has a type so I don't think we need another kind/type. And I think we should create guards for this instead in `public/app/features/variables/guard.ts`.

Reply:
`kind` is the important discriminator to distinguish between the different types of extended. We need some way to (statically) determine which kind of ExtendedState it is.

Importantly, `kind` is not a string, but is a union of string constants, so we get pretty strong type safety around them.

```typescript
typeof<kind> === "adhoc" | "query" | "datasource"
```

Also, it's [not possible to access incorrect properties](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIBMASB7BA1OKYOAIwBsIBRdYMbKAZTDkmQG8AoZbga1HQBcyAORx0AC1zCA3F26gY2ACoQAHmCEBnMERABzWd2ToWcBtgCuUJJqGooUOAE8APG2SR1WnaAPIAbnBkFhDeunrIAD7IIBZkZMgAvgB8hokcXKCQsIgoACKm5lZIBESkFNS09EwsKJxGfCCCIibMmpbWEDJyxoUdSEpOAA4QtmgOzm4eahrI2uHSAUEhYb5JqVzpmeDQ8EjIAIohUE6lxORUNHSMzKz1vPxCwgCOx07dRmflEEfQTpXXIQWEA8EDYADuIEM3FaZn6oTmPn0URicTIaQyHmGKC+FwB9Eo6ggIE0wGwIGQAF40FhcLiKldqrcUNECswip16ZcqjdaijficufjeZBZBwYMCEGAyRSoMT0NAABQzeUQZpCxlQQmQEkygCU7ExRgA9MbkCqmmrkMBNOA5iMEMAYMAEOawRY9OIADRzbDIBBwEBgsDIRA2TTIIZQbAjKDS0bIcnWsA9f3k9oUAB0ZGwemVRMt6EzCmUMz1YqM1pgyHzOoVRca6CplOpACJXn9Wwb7pXuKbzQX68hxHAI2DwcgSBBiTFCNHwVa6Ic3hqeTVID72sgF-7A8HQwhwx5xCgMJJXVGY9B45pU5WEOnsFmc3mLfXiyBFCp1OW70YHyST4QNmua1qqRawhyNi-pW6SJEAA) for the type.

I understand what you say about the variable already having a type, but due to the structure of the state (that type isn't in the Editor state) it can't be used to type these extended properties. The state could (theoretically) be rearranged to put everything together and we could make a single interface for each type of variable, but that all seems like a lot more work.

Another option could be use [`property in object` style narrowing](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIBMASB7BA1OKYOAIwBsIBRdYMbKAZTDkmQG8AoZb0GbAFQgAPMAC5kAZzBEQAcwDcXbuhZwG2AK5QkE8aihQ4ATwA8bZJBHipMhcgBucMhojXpoWcgA+yEBrJkyAC+AHyKwRxcoJCwiCgAIqrqWkgERKQU1LT0TCwonNzIKszJ2hD8RgAOELpoBsZmFsJiku5y8g5OLm62wWFcQZHI0dDwSMgAii5QRmnE5FQ0dIzMrAXccxkQU9BGWcviGiAA1iDYAO4g4cpJmmU9Ht6+-mThg1xgVSibC-v0lCIICAJMBsCBkABeNBYXA-TJLHKrFA+RIlO6pQjzeHZFZ5J47GZwxY43KQRQcGBHBBgUHgqBA9DQAAUzQZEHQ4iJfygAMgwNpAEp2ENCgB6UXIVkgRnoYYScCSaoIYAwYAISVnDSyAAWABpJNhkAg4CAzmBkIgdBJkJUoNhqlAaTVkGDhuajBAwEpuAgwRJsBQAHRkbCyFmA6XswO8ATNAXkwrDGDIJkAImKanRNVTw3BUplQvWie44slEZlyG1cGtZ3OyBIECBvkIdvO7IshowOHwmK23NJEG9xd9wIDEGDofDfJl0ZAfEEInjQ8TI-9QZDYfzUYzpR0S8TgyCQA) using properties that are unique to each subtype, but I prefer the method here - I think the explicitness makes it easier to understand what's going on.

Reply:
I'm also wanted to try and avoid user-defined type guards (`function isAdhoc(obj): obj is AdhocExtendedState`) as they're just another kind of "as" type cast/assertion.

Reply:
I don't agree with you, we have type safety today when we consume the state, but not when we dispatch actions. I don't agree with that the introduction of the `kind` will hinder you from dispatching the wrong editor action. I think the property in object style narrowing will suffice for this.